more_itertools/more.py

```diff
@@ -1634,35 +1634,148 @@ def numeric_range(*args):
         >>> start = datetime.datetime(2019, 1, 1)
         >>> stop = datetime.datetime(2019, 1, 3)
         >>> step = datetime.timedelta(days=1)
-        >>> items = numeric_range(start, stop, step)
+        >>> items = iter(numeric_range(start, stop, step))
         >>> next(items)
         datetime.datetime(2019, 1, 1, 0, 0)
         >>> next(items)
         datetime.datetime(2019, 1, 2, 0, 0)
 
     """
-    argc = len(args)
-    if argc == 1:
-        stop, = args
-        start = type(stop)(0)
-        step = 1
-    elif argc == 2:
-        start, stop = args
-        step = 1
-    elif argc == 3:
-        start, stop, step = args
-    else:
-        err_msg = 'numeric_range takes at most 3 arguments, got {}'
-        raise TypeError(err_msg.format(argc))
-
-    values = (start + (step * n) for n in count())
-    zero = type(step)(0)
-    if step > zero:
-        return takewhile(partial(gt, stop), values)
-    elif step < zero:
-        return takewhile(partial(lt, stop), values)
-    else:
-        raise ValueError('numeric_range arg 3 must not be zero')
+    _EMPTY_HASH = hash(range(0, 0))
+
+    def __init__(self, *args):
+        argc = len(args)
+        if argc == 1:
+            self._stop, = args
+            self._start = type(self._stop)(0)
+            self._step = 1
+        elif argc == 2:
+            self._start, self._stop = args
+            self._step = 1
+        elif argc == 3:
+            self._start, self._stop, self._step = args
+        else:
+            if argc == 0:
```

Comment:
Could you just move this into the `if / elif` structure above?

Reply:
You mean something like:

```
    elif argc == 0:
        raise TypeError('numeric_range expected at least 1 argument, got {}'.format(argc)
    else:
        raise TypeError('numeric_range expected at most 3 arguments, got {}'.format(argc)
```
?

(I added "at least" to make the message clearer).

Reply:
Yeah, exactly.

Reply:
Done.